# Exploratory Data Analysis

Just exploring a small part of the data to get insight into what I want to find as an anomaly


All data has been scraped from the web

In [2]:
#Libraries used for EDA
import numpy as np
import matplotlib.pyplot as plt
from scipy.io import loadmat
import pandas as pd

In [3]:
#Exploring just a small set of marathon data.
df = pd.read_csv("4male_result.csv") #encoding="ISO-8859-1")  #To deal with encoding issues

In [4]:
#Seeing first 5 rows of data
df.head()

Unnamed: 0                       Name    BIB Age Group  Age  Place gender  \
0           0           Turk, Matt (USA)  39595     25-29   29         15008   
1           0         Sugden, John (USA)  23526     45-49   46         15009   
2           0  Wojcikowski, Michal (USA)  21676     30-34   32         15010   
3           0        Leffler, Dale (USA)  32279     45-49   49         15011   
4           0      Andrade, Dalton (BRA)  45113     40-44   43         15012   

   Place Age  Place Overall Start Time Finish Time  ...  40K 40k Time of Day  \
0       1729          23214   08:14:49    04:23:08  ...  40K      12:25:57PM   
1       2176          23216   08:18:21    04:23:09  ...  40K      12:28:50PM   
2       2436          23217   07:45:22    04:23:09  ...  40K      11:53:36AM   
3       2177          23219   08:05:46    04:23:09  ...  40K      12:12:39PM   
4       2523          23221   08:42:38    04:23:09  ...  40K      12:49:50PM   

   40k Time 40k Difference 40k min/mile  Finish Finish Time of Day  \
0  04:11:08          31:55        10:16  Finish         12:37:57PM   
1  04:10:30          30:52        09:56  Finish         12:41:29PM   
2  04:08:14          38:38        12:26  Finish         12:08:31PM   
3  04:06:54          37:24        12:02  Finish         12:28:54PM   
4  04:07:12          37:45        12:09  Finish         01:05:47PM   

  Finish Time.1 Finish Difference Finish min/mile  
0      04:23:08              12:0           08:48  
1      04:23:09             12:39           09:17  
2      04:23:09             14:55           10:56  
3      04:23:09             16:15           11:55  
4      04:23:09             15:58           11:42  

[5 rows x 62 columns]

Get all the data types for the scraped data

In [5]:
df.dtypes

Unnamed: 0             int64
Name                  object
BIB                    int64
Age Group             object
Age                    int64
Place gender           int64
Place Age              int64
Place Overall          int64
Start Time            object
Finish Time           object
5k                    object
5k Time of Day        object
5k Time               object
5k Difference         object
5k min/mile           object
5k miles/h            object
10k                   object
10k Time of Day       object
10k Time              object
10k Difference        object
10k min/mile          object
10k miles/h           object
15K                   object
15k Time of Day       object
15k Time              object
15k Difference        object
15k min/mile          object
20K                   object
20k Time of Day       object
20k Time              object
                       ...  
Half                  object
Half Time of Day      object
Half Time             object
Half Differenc

Most of data is in object form need to convert to datetime.time.    
Thinking I don't necessarily need all columns.    
Wondering, which columns I need.  

In [6]:
from datetime import datetime
from dateutil.parser import parse
import pandas as pd

Getting just a dataframe of paces

In [7]:
pace = df[['Name','5k min/mile','10k min/mile','15k min/mile','20k min/mile','25k min/mile','30k min/mile','35k min/mile', '40k min/mile','Finish min/mile']]

In [8]:
pacelist = ['5k min/mile','10k min/mile','15k min/mile','20k min/mile','Half min/mile','25k min/mile','30k min/mile','35k min/mile','40k min/mile','Finish min/mile']

In [9]:
pacelist[0]

'5k min/mile'

In [10]:
type(pacelist)

list

In [11]:
pace.dtypes

Name               object
5k min/mile        object
10k min/mile       object
15k min/mile       object
20k min/mile       object
25k min/mile       object
30k min/mile       object
35k min/mile       object
40k min/mile       object
Finish min/mile    object
dtype: object

In [12]:
pace.to_csv(r'pace.csv')

In [13]:
testtime = "08:14:19"
dto = datetime.strptime(testtime, '%H:%M:%S').time()  #converted into time
print(dto)

08:14:19


In [14]:
df['5k min/mile']

0       09:50
1       11:01
2       08:58
3       09:00
4       08:41
5       09:29
6       09:31
7       09:03
8       09:36
9       09:43
10      08:34
11      08:59
12      08:34
13      09:33
14      08:53
15      09:30
16      09:14
17      09:30
18      09:44
19      09:05
20      08:25
21      08:23
22      09:55
23      07:58
24      08:55
25      09:56
26      08:55
27      09:27
28      09:25
29      09:05
        ...  
4929    10:22
4930    10:34
4931    09:35
4932    11:44
4933    10:48
4934    10:50
4935    10:50
4936    10:42
4937    09:32
4938    07:33
4939    10:57
4940    10:26
4941    09:54
4942    10:20
4943    08:48
4944    09:39
4945    10:55
4946    10:43
4947    09:06
4948    09:58
4949    11:54
4950    10:38
4951    10:20
4952    10:16
4953    09:20
4954    11:12
4955    10:12
4956    09:34
4957    10:00
4958    10:08
Name: 5k min/mile, Length: 4959, dtype: object

Making a function to take all the pace that are objects and makeing it into a datetime

In [22]:
example = pacelist[0]

s = pace[example].apply(lambda x: x.split(":"))
example2 = df['Start Time']

s2 = example2.apply(lambda x: x.split(":"))
from itertools import zip_longest

time = pd.DataFrame.from_records(zip_longest(*s.values)).T
time.columns = ['minutes', 'seconds']
time = time.replace('–', '-999').astype(np.float)

# time2 = pd.DataFrame.from_records(zip_longest(*s.values)).T
# time2.columns = ['hours', 'minutes','seconds']
# time2 = time2.replace('–', '-999').astype(np.float)


time.head()

minutes  seconds
0      9.0     50.0
1     11.0      1.0
2      8.0     58.0
3      9.0      0.0
4      8.0     41.0

In [23]:
print(s2)

0       [08, 14, 49]
1       [08, 18, 21]
2       [07, 45, 22]
3       [08, 05, 46]
4       [08, 42, 38]
5       [07, 39, 10]
6       [08, 14, 56]
7       [08, 39, 35]
8       [08, 16, 52]
9       [08, 45, 51]
10      [08, 09, 27]
11      [08, 05, 06]
12      [07, 51, 24]
13      [08, 38, 02]
14      [07, 52, 26]
15      [08, 35, 27]
16      [07, 52, 47]
17      [08, 35, 26]
18      [08, 06, 26]
19      [07, 49, 36]
20      [07, 34, 57]
21      [08, 38, 56]
22      [08, 35, 26]
23      [08, 35, 47]
24      [07, 44, 17]
25      [08, 12, 34]
26      [07, 44, 17]
27      [08, 14, 28]
28      [08, 35, 39]
29      [08, 11, 45]
            ...     
4929    [08, 44, 08]
4930    [08, 43, 30]
4931    [08, 04, 57]
4932    [08, 41, 51]
4933    [08, 42, 05]
4934    [08, 43, 51]
4935    [08, 46, 19]
4936    [08, 56, 33]
4937    [08, 06, 44]
4938    [08, 04, 31]
4939    [08, 38, 52]
4940    [08, 41, 55]
4941    [08, 50, 39]
4942    [08, 52, 37]
4943    [08, 02, 58]
4944    [08, 15, 11]
4945    [08, 

Function to convert pace

In [14]:
def paceconversion(x):
    for i in x:
#         pace[i]= pace[i].str.replace(':','.')
#         pace[i].convert_objects(convert_numeric=True)
        parts = np.array(pace[i].str.split(":").tolist())
        print(len(parts) == len(pace.index))
        minutes = pd.Series(parts[:,0]).astype(np.float).values
        seconds = pd.Series(parts[:,1]).astype(np.float).values
        pace[i] = minutes + seconds / 60

In [37]:
#paceconversion(pacelist)

In [16]:
pace.head() #Works

Name 5k min/mile 10k min/mile 15k min/mile  \
0           Turk, Matt (USA)       09:50        10:02        09:38   
1         Sugden, John (USA)       11:01        10:04        09:42   
2  Wojcikowski, Michal (USA)       08:58        09:03        09:11   
3        Leffler, Dale (USA)       09:00        09:09        09:12   
4      Andrade, Dalton (BRA)       08:41        08:39        08:35   

  20k min/mile 25k min/mile 30k min/mile 35k min/mile 40k min/mile  \
0        09:46        10:08        10:41        10:32        10:16   
1        09:43        09:41        10:07        09:44        09:56   
2        09:14        09:34        09:50        11:44        12:26   
3        09:11        09:26        10:17        10:56        12:02   
4        09:00        10:01        11:15        11:28        12:09   

  Finish min/mile  
0           08:48  
1           09:17  
2           10:56  
3           11:55  
4           11:42

In [17]:
pace.dtypes  #Still objects

Name               object
5k min/mile        object
10k min/mile       object
15k min/mile       object
20k min/mile       object
25k min/mile       object
30k min/mile       object
35k min/mile       object
40k min/mile       object
Finish min/mile    object
dtype: object

In [18]:
pace.to_csv(r'pace.csv')

Decided to do a simple way to transform paces into float.  
Not correct way of doing things but allows me to play with data in tableau.  

In [19]:
df['5k min/mile'] = df['5k min/mile'].str.replace(':','.') #Replace : to .
df['5k min/mile'].convert_objects(convert_numeric=True)  #Turn to float

/Users/johnvictorifermo/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: FutureWarning: convert_objects is deprecated.  To re-infer data dtypes for object columns, use Series.infer_objects()
For all other conversions use the data-type specific converters pd.to_datetime, pd.to_timedelta and pd.to_numeric.
  


0        9.50
1       11.01
2        8.58
3        9.00
4        8.41
5        9.29
6        9.31
7        9.03
8        9.36
9        9.43
10       8.34
11       8.59
12       8.34
13       9.33
14       8.53
15       9.30
16       9.14
17       9.30
18       9.44
19       9.05
20       8.25
21       8.23
22       9.55
23       7.58
24       8.55
25       9.56
26       8.55
27       9.27
28       9.25
29       9.05
        ...  
4929    10.22
4930    10.34
4931     9.35
4932    11.44
4933    10.48
4934    10.50
4935    10.50
4936    10.42
4937     9.32
4938     7.33
4939    10.57
4940    10.26
4941     9.54
4942    10.20
4943     8.48
4944     9.39
4945    10.55
4946    10.43
4947     9.06
4948     9.58
4949    11.54
4950    10.38
4951    10.20
4952    10.16
4953     9.20
4954    11.12
4955    10.12
4956     9.34
4957    10.00
4958    10.08
Name: 5k min/mile, Length: 4959, dtype: float64

In [20]:
#Getting Finish Time column into a datetime format
df['Finish Time'] = pd.to_datetime(df['Finish Time'],format= '%H:%M:%S' ).dt.time

In [21]:
# Seeing if it fixed objects into datetime
df.dtypes

Unnamed: 0             int64
Name                  object
BIB                    int64
Age Group             object
Age                    int64
Place gender           int64
Place Age              int64
Place Overall          int64
Start Time            object
Finish Time           object
5k                    object
5k Time of Day        object
5k Time               object
5k Difference         object
5k min/mile           object
5k miles/h            object
10k                   object
10k Time of Day       object
10k Time              object
10k Difference        object
10k min/mile          object
10k miles/h           object
15K                   object
15k Time of Day       object
15k Time              object
15k Difference        object
15k min/mile          object
20K                   object
20k Time of Day       object
20k Time              object
                       ...  
Half                  object
Half Time of Day      object
Half Time             object
Half Differenc

In [22]:
t = datetime.now().time()
seconds = (dto.hour * 60 + dto.minute) * 60 + dto.second  #To get time string into seconds

In [23]:
print(seconds)

29659


In [24]:
(8*60*60)+(14*60)+49  #Testing if accurate

29689

In [25]:
40.50/3.1  #Time diff divided by miles

13.064516129032258

In [26]:
x = 45.40/3.1  #min mile seems mostly accurate
print(x)

14.64516129032258


In [27]:
print(round(x, 2))

14.65


Will add temperature through day column to improve data.   
Thinking of currently using only pace min/mile as a good indicator.
Or take Difference between difference through seconds to show plus and negative at each 5k split.  
Will need to add half and 25k to get accurate 25k split.  

In [28]:
print(df.isnull().sum())

Unnamed: 0            0
Name                  0
BIB                   0
Age Group             0
Age                   0
Place gender          0
Place Age             0
Place Overall         0
Start Time            0
Finish Time           0
5k                    0
5k Time of Day        0
5k Time               0
5k Difference         0
5k min/mile           0
5k miles/h            0
10k                   0
10k Time of Day       0
10k Time              0
10k Difference        0
10k min/mile          0
10k miles/h           0
15K                   0
15k Time of Day       0
15k Time              0
15k Difference        0
15k min/mile          0
20K                   0
20k Time of Day       0
20k Time              0
                     ..
Half                  0
Half Time of Day      0
Half Time             0
Half Difference       0
Half min/mile         0
25K                   0
25k Time of Day       0
25k Time              0
25k Difference        0
25k min/mile          0
30K             

In [29]:
print(df.isnull().values.sum())  #NO nulls

0


## Ideas for project

Analytics with context    
Blend the Data, Technology, Subject  Matter Expertise and Analytics in a consumable way to guide actions to
drive Operational Discipline

How to find cheaters look at pace between split points, review missed split points.  
Explain timing chips.  Timing chips are typically encased in a hard plastic ring and   
given to each runner in their race packet prior to the event. The morning of the race,   
marathoners tie the ring into their shoelaces. Meantime, antennas sheathed in wide,   
rubber mats are positioned along the course. When the marathoner steps on the rubber mat   
across the starting line, his individual chip is electronically recognized by the timing  
system, and his time begins [source: Sinha]. Likewise, any time he or she steps on one of   
the specialty mats along the course, a computer records the accumulated time to that point in the race.  

# Ideas to visualize
Geocoding the locations of split spots
Create a dashboard story for project.
Work on Powerbi
Infographic for project


Pick out a name and show graph of splits    
Difference between male and female  
Difference in age groups  
Use pace to do anomaly detection  
Gif or video of people in china cheating in marathon by taking shortcut  
Get a model prediction with the data but only include first 35k and see accuracy of model  
Failure rate for timing chips when crossing mat  
Clusters
Null values, missed mats
Get Average
Cheat/Not cheat clasification.   


start  41.8830344,-87.6158958,15.84  
5k     41.8896808,-87.6300552,17.4  
10k    41.9314115,-87.6371737,16.58  
15k    41.9328129,-87.6397928,16.5  
20k    41.8918869,-87.6351928,17  
25k    41.8829413,-87.6524877,15  
30k    41.8617863,-87.6588821,15  
35k    41.8505029,-87.6426171,15  
40k    41.8505029,-87.6426171,15  
finish 41.8725803,-87.6251964,15  


In [30]:
#Geocoded data in case I want to map the Chicago marathon route
data = [['start', 41.8830344,-87.6158958],['5k', 41.8896808, -87.6300552], ['10k', 41.9314115,-87.6371737],['15k', 41.9328129,-87.6397928],
['20k', 41.8918869, -87.6351928],['25k', 41.8829413,-87.6524877],['30k', 41.8617863,-87.6588821], ['35k', 41.8505029,-87.6426171],['40k', 41.8505029,-87.6426171],
['42k', 41.8725803,-87.6251964]]




In [31]:
# Create the pandas DataFrame 
dfgeo = pd.DataFrame(data, columns = ['Distance', 'Lat', 'Long']) 

In [32]:
dfgeo

Distance        Lat       Long
0    start  41.883034 -87.615896
1       5k  41.889681 -87.630055
2      10k  41.931412 -87.637174
3      15k  41.932813 -87.639793
4      20k  41.891887 -87.635193
5      25k  41.882941 -87.652488
6      30k  41.861786 -87.658882
7      35k  41.850503 -87.642617
8      40k  41.850503 -87.642617
9      42k  41.872580 -87.625196

In [33]:
dfgeo.to_csv(r'geocoded_chi.csv')

In [34]:
ls

4male_result.csv*           Preprocess_data.ipynb
Anomaly_prediction.ipynb    Scraping_chicago_marathon/
Clustering_Learning.ipynb   full_uneconded_results.csv
Combine_datasets/           fullpace_uneconded.csv
Final_project.ipynb         geocoded_chi.csv
Images/                     pace.csv


In [39]:
#Quick exploratory of data
uedf = pd.read_csv("full_uneconded_results.csv")

In [40]:
uedf.head()

Unnamed: 0                     Name  BIB Age Group Age  Place gender  \
0           0  Cherono, Lawrence (KEN)    4     30-34  31             1   
1           1     Debela, Dejene (ETH)   38     20-24  24             2   
2           2     Mengstu, Asefa (ETH)    5     30-34  31             3   
3           3      Karoki, Bedan (KEN)    9     25-29  29             4   
4           4       Abdi, Bashir (BEL)   10     30-34  30             5   

  Place Age  Place Overall Start Time Finish Time  ... 40k Time of Day  \
0         1              1   07:30:03    02:05:45  ...      09:29:12AM   
1         1              2   07:30:03    02:05:46  ...      09:29:10AM   
2         2              3   07:30:03    02:05:48  ...      09:29:10AM   
3         1              4   07:30:03    02:05:53  ...      09:29:11AM   
4         3              5   07:30:03    02:06:14  ...      09:29:55AM   

   40k Time 40k Difference 40k min/mile  Finish Finish Time of Day  \
0  01:59:10          15:17        04:55  Finish         09:35:47AM   
1  01:59:08          15:15        04:55  Finish         09:35:48AM   
2  01:59:08          15:15        04:55  Finish         09:35:50AM   
3  01:59:09          15:16        04:55  Finish         09:35:55AM   
4  01:59:53          14:39        04:43  Finish         09:36:16AM   

  Finish Time.1 Finish Difference Finish min/mile sex  
0      02:05:45             06:36           04:50   M  
1      02:05:46             06:39           04:52   M  
2      02:05:48             06:40           04:53   M  
3      02:05:53             06:45           04:57   M  
4      02:06:14             06:21           04:40   M  

[5 rows x 63 columns]

In [1]:
#Dropping unecessary columns
#uedf = uedf.drop(columns =['Unnamed: 0'])

In [43]:
paceue = uedf[['Name','sex', 'Start Time','Finish Time of Day','5k min/mile','10k min/mile','15k min/mile','20k min/mile', 'Half min/mile','25k min/mile','30k min/mile','35k min/mile', '40k min/mile','Finish min/mile']]

In [44]:
paceue.head()

Name sex Start Time Finish Time of Day 5k min/mile  \
0  Cherono, Lawrence (KEN)   M   07:30:03         09:35:47AM       04:45   
1     Debela, Dejene (ETH)   M   07:30:03         09:35:48AM       04:45   
2     Mengstu, Asefa (ETH)   M   07:30:03         09:35:50AM       04:45   
3      Karoki, Bedan (KEN)   M   07:30:03         09:35:55AM       04:45   
4       Abdi, Bashir (BEL)   M   07:30:03         09:36:16AM       04:46   

  10k min/mile 15k min/mile 20k min/mile Half min/mile 25k min/mile  \
0        04:45        04:45        04:47         04:45        04:50   
1        04:45        04:45        04:47         04:44        04:49   
2        04:45        04:45        04:47         04:44        04:49   
3        04:44        04:45        04:47         04:44        04:49   
4        04:44        04:48        04:53         04:57        04:57   

  30k min/mile 35k min/mile 40k min/mile Finish min/mile  
0        04:51        04:48        04:55           04:50  
1        04:51        04:48        04:55           04:52  
2        04:51        04:48        04:55           04:53  
3        04:52        04:48        04:55           04:57  
4        05:00        04:48        04:43           04:40

In [47]:
#Paceconversion function
def paceueconversion(x):
    for i in x:
        paceue[i]= paceue[i].str.replace(':','.')
        #pace[i].convert_objects(convert_numeric=True)

In [48]:
paceueconversion(pacelist)

/Users/johnvictorifermo/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [49]:
paceue.head()

Name sex Start Time Finish Time of Day 5k min/mile  \
0  Cherono, Lawrence (KEN)   M   07:30:03         09:35:47AM       04.45   
1     Debela, Dejene (ETH)   M   07:30:03         09:35:48AM       04.45   
2     Mengstu, Asefa (ETH)   M   07:30:03         09:35:50AM       04.45   
3      Karoki, Bedan (KEN)   M   07:30:03         09:35:55AM       04.45   
4       Abdi, Bashir (BEL)   M   07:30:03         09:36:16AM       04.46   

  10k min/mile 15k min/mile 20k min/mile Half min/mile 25k min/mile  \
0        04.45        04.45        04.47         04.45        04.50   
1        04.45        04.45        04.47         04.44        04.49   
2        04.45        04.45        04.47         04.44        04.49   
3        04.44        04.45        04.47         04.44        04.49   
4        04.44        04.48        04.53         04.57        04.57   

  30k min/mile 35k min/mile 40k min/mile Finish min/mile  
0        04.51        04.48        04.55           04.50  
1        04.51        04.48        04.55           04.52  
2        04.51        04.48        04.55           04.53  
3        04.52        04.48        04.55           04.57  
4        05.00        04.48        04.43           04.40

In [50]:
paceue.to_csv(r'fullpace_uneconded.csv')

In [51]:
paceue.dtypes

Name                  object
sex                   object
Start Time            object
Finish Time of Day    object
5k min/mile           object
10k min/mile          object
15k min/mile          object
20k min/mile          object
Half min/mile         object
25k min/mile          object
30k min/mile          object
35k min/mile          object
40k min/mile          object
Finish min/mile       object
dtype: object

In [52]:
paceue.isna().sum()

Name                  0
sex                   0
Start Time            0
Finish Time of Day    0
5k min/mile           0
10k min/mile          0
15k min/mile          0
20k min/mile          0
Half min/mile         0
25k min/mile          0
30k min/mile          0
35k min/mile          0
40k min/mile          0
Finish min/mile       0
dtype: int64

## Trying to find possible bib mules in data.  
Bib mules are those that have will carry someone elses big across the finish line along with their own bib.  
Should find people with same timing splits in data.  

In [56]:
#Find possible bib mules
# Select all duplicate rows based on one column
nameRowsDF = paceue[paceue.duplicated(['Name'])]
print("Duplicate Rows based on 1 columns are:", nameRowsDF, sep='\n')


Duplicate Rows based on 1 columns are:
                               Name sex Start Time Finish Time of Day  \
1034             Walsh, Brian (USA)   M   07:31:52         10:26:38AM   
2852           Garcia, Miguel (USA)   M   07:32:24         10:43:01AM   
3690               Lang, Paul (USA)   M   07:35:22         10:52:24AM   
3859                Wang, Hao (CHN)   M   07:33:57         10:52:20AM   
3981                Kim, Mark (USA)   M   07:34:20         10:53:37AM   
4026              Jiang, Jing (CHN)   M   07:47:50         11:07:25AM   
4191           Murphy, Andrew (USA)   M   07:47:43         11:08:34AM   
4308              Bailey, Dan (USA)   M   07:34:22         10:55:56AM   
4707           Mcgrath, David (USA)   M   07:37:42         11:01:49AM   
4857           Mcgrath, David (USA)   M   07:38:31         11:03:31AM   
4858              Smith, Matt (USA)   M   07:33:50         10:58:50AM   
4951            James, Andrew (USA)   M   07:48:51         11:14:21AM   
5237        

In [57]:
nameRowsDF.shape

(399, 14)

In [59]:
#Find possible bib mules
# Select all duplicate rows based on one column
duponeRowsDF = paceue[paceue.duplicated(['Finish Time of Day'])]
print("Duplicate Rows based on 1 columns are:", duponeRowsDF, sep='\n')


Duplicate Rows based on 1 columns are:
                                    Name sex Start Time Finish Time of Day  \
86                     Leys, Dries (BEL)   M   07:30:03         09:58:58AM   
100          Biebelhausen, Stephen (USA)   M   07:30:03         10:00:51AM   
127                 Benard, Nathan (FRA)   M   07:31:14         10:03:56AM   
161        Swiatkowski, Mieczyslaw (USA)   M   07:31:17         10:06:35AM   
170                 Thoburn, Chris (USA)   M   07:30:03         10:06:04AM   
190                Pavelko, Andrii (UKR)   M   07:31:14         10:08:31AM   
203               Bareikis, Arturs (USA)   M   07:30:03         10:07:43AM   
208                Mulvihill, John (IRL)   M   07:31:20         10:09:07AM   
209                    Spak, Aaron (USA)   M   07:31:15         10:09:07AM   
215                Tarbath, Forest (USA)   M   07:30:03         10:08:00AM   
228               Talleri, William (ITA)   M   07:31:20         10:10:07AM   
239                Ernst,

In [60]:
duponeRowsDF.shape

(28117, 14)

Based on duplicates people finishing at the same time of day I got 28,117 results way too many.

In [61]:
#Find possible bib mules.  People carrying someone elses bib to get them a boston qualifier
# Select all duplicate rows based on two columns
duptwoRowsDF = paceue[paceue.duplicated(['Start Time','Finish Time of Day'])]
print("Duplicate Rows based on a 2 columns are:", duptwoRowsDF, sep='\n')


Duplicate Rows based on a 2 columns are:
                                     Name sex Start Time Finish Time of Day  \
215                 Tarbath, Forest (USA)   M   07:30:03         10:08:00AM   
381                  Umhau, Bennett (USA)   M   07:31:19         10:15:19AM   
500                  Silva, Marcelo (BRA)   M   07:31:48         10:18:41AM   
542               Calliari, Michele (ITA)   M   07:31:23         10:19:12AM   
570                    Sutley, Ryan (USA)   M   07:31:18         10:19:39AM   
605          Bergesen, Morten Weber (NOR)   M   07:31:25         10:20:15AM   
726      Gonzalez Guerrero, Benigno (ESP)   M   07:31:30         10:22:24AM   
772    Ortega Marsiglia, Juan Pablo (CHI)   M   07:31:16         10:22:53AM   
823                     Mann, Shaun (GBR)   M   07:31:23         10:23:48AM   
843                   Rotundo, Mark (USA)   M   07:31:43         10:24:21AM   
951                      Cheng, Min (USA)   M   07:31:20         10:25:19AM   
1023       

In [62]:
duptwoRowsDF.shape

(1055, 14)

Based on Start time and finish time all the same I got 1,055 results.

In [64]:
paceue = paceue[paceue["Finish Time of Day"].duplicated(keep=False)]

a = paceue.groupby("Finish Time of Day").apply(lambda x: list(x.index))

In [65]:
print(a)  #Over 11,000 duplicate finish times.

Finish Time of Day
01:00:00PM                                       [17526, 31746]
01:00:01PM                  [18152, 18763, 19295, 20035, 37375]
01:00:02PM                         [14737, 35599, 35679, 37144]
01:00:03PM                  [13848, 20248, 31871, 34264, 34511]
01:00:04PM    [18537, 31536, 31709, 31762, 32422, 34775, 353...
01:00:05PM                                       [18175, 32261]
01:00:09PM                         [17844, 20984, 32210, 36770]
01:00:10PM                         [17455, 30408, 34334, 34335]
01:00:12PM                                       [32167, 38214]
01:00:13PM                                       [14526, 19098]
01:00:17PM           [18969, 19247, 31011, 35630, 36146, 36229]
01:00:18PM                                       [32391, 36143]
01:00:19PM                         [20071, 32362, 38409, 38494]
01:00:20PM                                [16759, 17489, 31763]
01:00:21PM                                [12548, 17798, 31098]
01:00:22PM           

In [66]:
bibmule = paceue[paceue.duplicated(subset=['Start Time','Finish Time of Day'], keep=False)]
len(bibmule)

2075

In [67]:
print(bibmule)   #Able to get over 2000 results using 2 columns for bibmules.  Need another filter.

                                     Name sex Start Time Finish Time of Day  \
94                 Johnson, Spencer (USA)   M   07:30:03         09:59:47AM   
166                  Butcher, Brian (USA)   M   07:30:03         10:05:43AM   
214                    Bell, Samuel (USA)   M   07:30:03         10:08:00AM   
215                 Tarbath, Forest (USA)   M   07:30:03         10:08:00AM   
379                 Diviney, Ciaran (IRL)   M   07:31:19         10:15:19AM   
381                  Umhau, Bennett (USA)   M   07:31:19         10:15:19AM   
451               Baddiley, Nicolas (AUS)   M   07:30:03         10:15:40AM   
499            Evangelista, Gustavo (BRA)   M   07:31:48         10:18:41AM   
500                  Silva, Marcelo (BRA)   M   07:31:48         10:18:41AM   
541                       Ho, David (SWE)   M   07:31:23         10:19:12AM   
542               Calliari, Michele (ITA)   M   07:31:23         10:19:12AM   
569                   Grimsbu, Mads (NOR)   M   07:3

In [68]:
bibmule = paceue[paceue.duplicated(subset=['Start Time','Finish Time of Day', '5k min/mile', '10k min/mile', '15k min/mile', '20k min/mile', 'Half min/mile','25k min/mile', '30k min/mile', '35k min/mile', '40k min/mile'], keep=False)]
len(bibmule)

310

With more features was able to narrow it down to almost 300 Hard to narrow down lower without looking through all the pictures

In [69]:
print(bibmule)  #With more features was able to narrow it down to almost 300 Hard to narrow down lower without looking through all the pictures

                                  Name sex Start Time Finish Time of Day  \
1354               Smith, Nelson (USA)   M   07:34:07         10:31:49AM   
1355                Rushek, Mark (USA)   M   07:34:07         10:31:49AM   
1535                  Ding, Zuyu (CHN)   M   07:31:51         10:30:47AM   
1537                Zhu, Yongbin (CHN)   M   07:31:51         10:30:47AM   
2264               Schuster, Tim (USA)   M   07:34:42         10:40:19AM   
2423             Lindblade, Kris (USA)   M   07:42:03         10:49:15AM   
3104              Oreglia, Diego (ITA)   M   07:38:25         10:51:04AM   
3541             Touloupas, John (USA)   M   07:41:35         10:57:29AM   
4646               Bene, Patrick (USA)   M   07:38:07         11:01:54AM   
4647                Bene, George (USA)   M   07:38:07         11:01:54AM   
4973          Balsalobre, Carlos (ESP)   M   07:42:47         11:08:23AM   
4974         Balsalobre, Ignacio (ESP)   M   07:42:47         11:08:23AM   
5206        

Getting names of all possible bibmules

In [71]:
list(bibmule['Name'])

['Smith, Nelson (USA)',
 'Rushek, Mark (USA)',
 'Ding, Zuyu (CHN)',
 'Zhu, Yongbin (CHN)',
 'Schuster, Tim (USA)',
 'Lindblade, Kris (USA)',
 'Oreglia, Diego (ITA)',
 'Touloupas, John (USA)',
 'Bene, Patrick (USA)',
 'Bene, George (USA)',
 'Balsalobre, Carlos (ESP)',
 'Balsalobre, Ignacio (ESP)',
 'Nichols, Matthew (USA)',
 'Toupin, Hugo (CAN)',
 'Ford, Michael (USA)',
 'Zenner, John (USA)',
 'Nyman, Charles (USA)',
 'Nyman, Andrew (GBR)',
 'Hakenes, Michael (GER)',
 'Niehus, Henrik (GER)',
 'Misek, Glen (USA)',
 'Brown, Michael (USA)',
 'Stegenga, Jorrit (NED)',
 'Hekman, Bas (NED)',
 'Connelly, Chris (USA)',
 'Aguilera, Jorge (USA)',
 'Ko, Matthew (USA)',
 'Flack, Logan (USA)',
 'Cole, Christopher (USA)',
 'Chavez, Jesus (MEX)',
 'Noonan, Eddie (USA)',
 'Domingo, Jose (ESP)',
 'Villa, Pedro (COL)',
 'Branch, Phillip (AFG)',
 'Maqueda, Javier (ESP)',
 'Torres Fink, Eduardo (ESP)',
 'De Rooij, Sebastiaan (NED)',
 'Niessen, Jan (NED)',
 'Fernandez, Amador (MEX)',
 'Booth, William (USA)'

## More ideas for capstone project

Explain Boston Marathon  
Significance  
Boston marathon qualifying times maybe just limit it to this.  
Marathon Stats
Taken from (Run Repeat website)[https://runrepeat.com/state-of-running]  

Running has grown by 57% over the past 10 years, which is great, but any sport, as it attracts the masses, will have a period of decline. We cannot say if this is a short-term trend or a long term one. Of course, the industry should be alert.

10 years ago, running a marathon was the ultimate goal for many athletes. Not many did it. It attracted a certain audience. Then a group of less experienced athletes joined the pack and proved that they could do it as well. The trend kept going, and at some point, the the extreme athletes did not find the marathon extreme anymore. They did not feel special, which was an important element of running the marathon. Because of this, the ultramarathon, trail running, the Ironman and so on exploded in popularity.

The motives for running are potentially changing, and the races have not picked up on this trend. There are quite a few indicators of this. In our report, we prove how 1) running at milestone ages (30, 40, 50, 60) is significantly less important in 2019 than it was 15 years ago, 2) people travel more than ever to race and 3) the finish times are slower. With people being slower, it's not the individual who's getting slower, but the average of all runners, meaning that the "demography" has changed. More slow runners participate. These three indicate a growth in the experience economy rather than achievement-motives. This is very important, we think, and we think that the industry hasn't been good enough at adjusting their races.  

Taken from (Bizjournals)[https://www.bizjournals.com/chicago/news/2019/10/02/bank-of-america-chicago-marathon-economic-impact.html]

The Bank of America Chicago Marathon said it had a record economic impact on the city last year.  

Marathon officials on Wednesday said the 2018 event contributed a record-breaking $378 million to the city’s economy, a $40-million increase over the 2017 race.  

The race’s 2018 economic impact marked the second year in a row that the total figure exceeded $300 million — positively impacting tourism, hospitality and a multitude of businesses throughout the city.

Taken from (Investopedia)[https://www.investopedia.com/articles/investing/100815/economics-behind-marathons.asp]

Marathons, surprisingly, are a relatively inelastic good. Registration fees are up, yet more people participate in running events each year. Some of the elite races, such as the New York City Marathon, are able to charge $11 for entry into a selection lottery just for a chance to participate in the actual marathon, which commands a whopping $255 registration fee.    

What Is Inelastic?  
Inelastic is an economic term referring to the static quantity of a good or service when its price changes. Inelastic means that when the price goes up, consumers’ buying habits stay about the same, and when the price goes down, consumers’ buying habits also remain unchanged.

The Bottom Line  
While modern marathon racing is over a century old, the participation numbers have increased wildly in the past decade. As more people embrace the healthy benefits of running, along with its low-cost and minimal barriers to entry, marathon organizers are shaping up to see their numbers continue to grow each year.  

# Boston Qualifying times

Male  
AGE GROUP	QUALIFYING STANDARD	QUALIFICATION TIMES ACCEPTED (faster than and including)  
18-34	3hrs 00min 00sec	2hrs 58min 21sec  
35-39	3hrs 05min 00sec	3hrs 03min 21sec  
40-44	3hrs 10min 00sec	3hrs 08min 21sec  
45-49	3hrs 20min 00sec	3hrs 18min 21sec  
50-54	3hrs 25min 00sec	3hrs 23min 21sec  
55-59	3hrs 35min 00sec	3hrs 33min 21sec  
60-64	3hrs 50min 00sec	3hrs 48min 21sec  
65-69	4hrs 05min 00sec	4hrs 03min 21sec  
70-74	4hrs 20min 00sec	4hrs 18min 21sec  
75-79	4hrs 35min 00sec	4hrs 33min 21sec  
80 and over	4hrs 50min 00sec	4hrs 48min 21sec  

Female  
AGE GROUP	QUALIFYING STANDARD	QUALIFICATION TIMES ACCEPTED (faster than and including)  
18-34	3hrs 30min 00sec	3hrs 28min 21sec  
35-39	3hrs 35min 00sec	3hrs 33min 21sec  
40-44	3hrs 40min 00sec	3hrs 38min 21sec  
45-49	3hrs 50min 00sec	3hrs 48min 21sec  
50-54	3hrs 55min 00sec	3hrs 53min 21sec  
55-59	4hrs 05min 00sec	4hrs 03min 21sec  
60-64	4hrs 20min 00sec	4hrs 18min 21sec  
65-69	4hrs 35min 00sec	4hrs 33min 21sec  
70-74	4hrs 50min 00sec	4hrs 48min 21sec  
75-79	5hrs 05min 00sec	5hrs 03min 21sec  
80 and over	5hrs 20min 00sec	5hrs 18min 21sec  


* Qualified entrants for the 2020 Boston Marathon must have met the designated time standard that corresponds with their age group and gender in a certified marathon on or after September 15, 2018. Proof of qualification must accompany each athlete's application, and participants are required to be 18 years or older on Boston Marathon race day.    

The qualifying window for the 2021 Boston Marathon began on Saturday, September 14, 2019, and will continue through the conclusion of 2021 Boston Marathon registration next fall. Registration dates for the 2021 Boston Marathon will be announced next year.     

Decided to limit it to times under 4:35 if someone over 70 wants to run Boston or 80 more power to him.  Now everyone else play fair.  
